modules/axtask/src/run_queue.rs

```diff
@@ -1,72 +1,249 @@
 use lazyinit::LazyInit;
 use scheduler::BaseScheduler;
 
+use axhal::cpu::this_cpu_id;
+
 use crate::task::{CurrentTask, TaskState};
 use crate::{AxTaskRef, Scheduler, TaskInner, WaitQueue};
 
-// TODO: per-CPU
-pub(crate) static RUN_QUEUE: LazyInit<SpinNoIrq<AxRunQueue>> = LazyInit::new();
+macro_rules! percpu_static {
+    ($($name:ident: $ty:ty = $init:expr),* $(,)?) => {
+        $(
+            #[percpu::def_percpu]
+            static $name: $ty = $init;
+        )*
+    };
+}
 
-// TODO: per-CPU
-static EXITED_TASKS: SpinNoIrq<VecDeque<AxTaskRef>> = SpinNoIrq::new(VecDeque::new());
+percpu_static! {
+    RUN_QUEUE: LazyInit<AxRunQueue> = LazyInit::new(),
+    EXITED_TASKS: VecDeque<AxTaskRef> = VecDeque::new(),
+    WAIT_FOR_EXIT: WaitQueue = WaitQueue::new(),
+    IDLE_TASK: LazyInit<AxTaskRef> = LazyInit::new(),
+}
+
+/// An array of references to run queues, one for each CPU, indexed by cpu_id.
+///
+/// This static variable holds references to the run queues for each CPU in the system.
+///
+/// # Safety
+///
+/// Access to this variable is marked as `unsafe` because it contains `MaybeUninit` references,
+/// which require careful handling to avoid undefined behavior. The array should be fully
+/// initialized before being accessed to ensure safe usage.
+static mut RUN_QUEUES: [MaybeUninit<&'static mut AxRunQueue>; axconfig::SMP] =
+    [ARRAY_REPEAT_VALUE; axconfig::SMP];
+const ARRAY_REPEAT_VALUE: MaybeUninit<&'static mut AxRunQueue> = MaybeUninit::uninit();
+
+/// Returns a reference to the current run queue.
+///
+/// ## Safety
+///
+/// This function returns a static reference to the current run queue, which
+/// is inherently unsafe. It assumes that the `RUN_QUEUE` has been properly
+/// initialized and is not accessed concurrently in a way that could cause
+/// data races or undefined behavior.
+///
+/// ## Returns
+///
+/// A static reference to the current run queue.
+// #[inline(always)]
+pub(crate) fn current_run_queue<G: BaseGuard>() -> AxRunQueueRef<'static, G> {
+    let irq_state = G::acquire();
+    AxRunQueueRef {
+        inner: unsafe { RUN_QUEUE.current_ref_mut_raw() },
+        state: irq_state,
+        _phantom: core::marker::PhantomData,
+    }
+}
 
-static WAIT_FOR_EXIT: WaitQueue = WaitQueue::new();
+/// Selects the run queue index based on a CPU set bitmap, minimizing the number of tasks.
+///
+/// This function filters the available run queues based on the provided `cpumask` and
+/// selects the one with the fewest tasks. The selected run queue's index (cpu_id) is returned.
+///
+/// ## Arguments
+///
+/// * `cpumask` - A bitmap representing the CPUs that are eligible for task execution.
+///
+/// ## Returns
+///
+/// The index (cpu_id) of the selected run queue.
+///
+/// ## Panics
+///
+/// This function will panic if there is no available run queue that matches the CPU set.
+///
+#[cfg(feature = "smp")]
+#[inline]
+fn select_run_queue_index(cpumask: CpuMask<{ axconfig::SMP }>) -> usize {
```

Comment:
Add a type alias:

```rs
type CpuMask = cpumask::CpuMask<{ axconfig::SMP }>;
```

Reply:
Can we get rid of the brackets around `axconfig::SMP`?

Reply:
```bash
error[E0573]: expected type, found constant `axconfig::SMP`
  --> /home/hky/workspace/arceos-org/arceos/modules/axtask/src/run_queue.rs:87:44
   |
87 | fn select_run_queue_index(cpumask: CpuMask<axconfig::SMP>) -> usize {
   |                                            ^^^^^^^^^^^^^ not a type

error[E0747]: unresolved item provided when a constant was expected
  --> /home/hky/workspace/arceos-org/arceos/modules/axtask/src/run_queue.rs:87:44
   |
87 | fn select_run_queue_index(cpumask: CpuMask<axconfig::SMP>) -> usize {
   |                                            ^^^^^^^^^^^^^
   |
help: if this generic argument was intended as a const parameter, surround it with braces
   |
87 | fn select_run_queue_index(cpumask: CpuMask<{ axconfig::SMP }>) -> usize {
   |                                            +               +

Some errors have detailed explanations: E0573, E0747.
For more information about an error, try `rustc --explain E0573`.
error: could not compile `axtask` (lib) due to 2 previous errors
make: *** [scripts/make/build.mk:34: _cargo_build] Error 101
```

Reply:
https://users.rust-lang.org/t/why-do-const-generics-need-braces-when-the-path-of-a-constant-is-composed-of/111240